### PR TITLE
Add managed memory allocator.

### PR DESCRIPTION
**Pull request by trivialfis**
For experimenting with external memory.

### CODE REVIEW

**Comment by trivialfis** [comment]
cc @rongou .

File: src/common/cuda_pinned_allocator.h [hunk 1]
```diff
@@ -1,93 +1,114 @@
...
 template <typename T>
-class pinned_allocator;
+struct PinnedAllocPolicy {
+  using pointer = T*;              // NOLINT: The type returned by address() / allocate()
+  using const_pointer = const T*;  // NOLINT: The type returned by address()
+  using size_type = std::size_t;   // NOLINT: The type used for the size of the allocation
+  using value_type = T;            // NOLINT: The type of the elements in the allocator
+
+  size_type max_size() const {  // NOLINT
+    return (std::numeric_limits<size_type>::max)() / sizeof(value_type);
```
Comment (rongou) [inline src/common/cuda_pinned_allocator.h:33]: Shouldn't it just be:
```cpp
return std::numeric_limits<size_type>::max() / sizeof(value_type);
```

File: src/common/cuda_pinned_allocator.h [hunk 2]
```diff
@@ -1,93 +1,114 @@
...
 template <typename T>
-class pinned_allocator {
+struct ManagedAllocPolicy {
+  using pointer = T*;              // NOLINT: The type returned by address() / allocate()
+  using const_pointer = const T*;  // NOLINT: The type returned by address()
+  using size_type = std::size_t;   // NOLINT: The type used for the size of the allocation
+  using value_type = T;            // NOLINT: The type of the elements in the allocator
+
+  size_type max_size() const {  // NOLINT
+    return (std::numeric_limits<size_type>::max)() / sizeof(value_type);
```
Comment (rongou) [inline src/common/cuda_pinned_allocator.h:57]: ditto

**Review by rongou — APPROVED** [review]

File: src/common/cuda_pinned_allocator.h (hunk 1 above)
Comment (trivialfis) [reply src/common/cuda_pinned_allocator.h:33]: Yeah, that's from the old code, haven't done the cleanup yet, will change it.

File: src/common/cuda_pinned_allocator.h (hunk 2 above)
Comment (trivialfis) [reply src/common/cuda_pinned_allocator.h]: Done.

**Pull request merged**